internal/overlord/logstate/loki.go

```diff
@@ -0,0 +1,162 @@
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logstate
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/servicelog"
+)
+
+const (
+	requestStart = `
+{
+  "streams": [
+    {
+      "stream": {
+        "service": %q
+      },
+      "values": [`
+
+	requestEnd = `
+      ]
+    }
+  ]
+}`
+
+	lokiPushAPI = "/loki/api/v1/push"
+)
+
+// lokiClient implements logClient
+// We send a new request to Loki when either:
+//   - we have maxMessages ready to send; or
+//   - delay has passed since the last request.
+type lokiClient struct {
+	service     string
+	lokiURL     *url.URL
+	maxMessages int
+	delay       time.Duration
+
+	// buffered request so far
+	buffer *bytes.Buffer
+	// how many messages have been written so far
+	messages int
+	// write deadline for flushing buffer
+	deadline *time.Time
+}
+
+func newLokiClient(service, url string, maxMessages int, delay time.Duration) (*lokiClient, error) {
+	lokiURL, err := parseURL(url)
+	if err != nil {
+		return nil, err
+	}
+	c := &lokiClient{
+		service:     service,
+		lokiURL:     lokiURL,
+		maxMessages: maxMessages,
+		delay:       delay,
+	}
+	c.reset()
+	return c, nil
+}
+
+func parseURL(rawURL string) (*url.URL, error) {
+	u, err := url.Parse(rawURL)
+	if err != nil || u.Host == "" {
+		u, err = url.Parse("//" + rawURL)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("invalid Loki server address: %v", err)
+	}
+
+	// check for and set loki defaults if omitted from address
+	if u.Scheme == "" {
+		u.Scheme = "http"
+		if u.Port() == "" {
+			u.Host += ":3100"
+		}
+	}
+
+	if u.Scheme != "http" { // https??
+		return nil, fmt.Errorf("unsupported Loki URL scheme '%v'", u.Scheme)
+	}
+
+	return u, nil
+}
+
+func (c *lokiClient) reset() {
+	c.buffer = bytes.NewBuffer([]byte(fmt.Sprintf(requestStart, c.service)))
+	c.messages = 0
+	c.deadline = nil
+}
+
+// Check if it is time to send a request to remote.
+func (c *lokiClient) MaybeFlush() {
+	if c.messages >= c.maxMessages {
+		c.Flush()
+	}
+	if c.deadline != nil && time.Now().After(*c.deadline) {
+		c.Flush()
+	}
+}
+
+// send currently written request to the remote Loki
+// and reset state
+func (c *lokiClient) Flush() {
+	c.buffer.WriteString(requestEnd)
+	//logger.Noticef(c.buffer.String())
+	resp, err := http.Post(c.lokiURL.String()+lokiPushAPI, "application/json", c.buffer)
+	if err != nil {
+		logger.Noticef("error sending HTTP request: %v", err)
```

Comment:
`%w` only makes sense for `fmt.Errorf`

Reply:
fair, I can add more context to the message

Reply:
Yep, `%w` isn't meaningful for logs.

Unrelated nit: we've settled on the word "cannot" rather than "error" for these things in Pebble. Also, we capitalize log messages (but not error messages, as they may be in a "colon chain"). So this would be something like:

```go
logger.Noticef("Cannot send log to Loki: %v", err)
```